src/main/java/org/keycloak/protocol/cas/endpoints/AbstractValidateEndpoint.java

```diff
@@ -74,52 +84,85 @@ protected void checkClient(String service) {
         session.getContext().setClient(client);
     }
 
-    protected void checkTicket(String ticket, boolean requireReauth) {
+    protected void checkTicket(String ticket, String prefix, boolean requireReauth) {
         if (ticket == null) {
             event.error(Errors.INVALID_CODE);
             throw new CASValidationException(CASErrorCode.INVALID_REQUEST, "Missing parameter: " + CASLoginProtocol.TICKET_PARAM, Response.Status.BAD_REQUEST);
         }
-        if (!ticket.startsWith(CASLoginProtocol.SERVICE_TICKET_PREFIX)) {
+
+        if (!ticket.startsWith(prefix)) {
             event.error(Errors.INVALID_CODE);
             throw new CASValidationException(CASErrorCode.INVALID_TICKET_SPEC, "Malformed service ticket", Response.Status.BAD_REQUEST);
         }
```

Comment:
Hi @jacekkow , regarding the example above, I verified that each endpoint validates the expected tickets, /serviceValidate and /validate only ST-, /proxy only PGT- and /proxyValidate ST- and PT-.
 Anything other than the "prefix" parameter generates an invalid ticket.

Reply:
That's not what I'm worried about - I'm worried that you can get valid "ST-" ticket, replace "ST-" with "PGT-" and use such ST-changed-into-PGT token to receive "PT-". I don't see the ticket type being persisted server-side nor checked for "misprefixing".

Reply:
Yes you are right. I added the prefix to the storage key. This way, "misprefixing" does not find the ticket on the server-side.